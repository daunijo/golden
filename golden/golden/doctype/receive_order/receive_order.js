// Copyright (c) 2018, RSS and contributors
// For license information, please see license.txt

frappe.ui.form.on('Receive Order', {
	setup: function(frm){
		frm.set_query('accepted_location', function(doc) {
			return {
				filters: { 'is_group': 0 }
			}
		});
		frm.set_query("expedition", function (doc) {
			return {
				filters: {
					'buying': 1
				}
			}
		});
		frm.set_indicator_formatter('item_code',
			function(doc) {
				return (doc.qty == doc.si_qty) ? "green" : "blue"
			}
		)
	},
	onload: function(frm) {
		frm.set_query("item_code", "items", function(doc, cdt, cdn) {
			var row = locals[cdt][cdn];
			return {
				query: "golden.golden.doctype.receive_order.receive_order.get_item_code",
			}
		});
		frm.set_query("purchase_order", "items", function(doc, cdt, cdn) {
			var row = locals[cdt][cdn];
			return {
				query: "golden.golden.doctype.receive_order.receive_order.get_list_purchase_order",
				filters: {
					'item_code': row.item_code,
				}
			}
		});
		frm.set_query("uom", "items", function(doc, cdt, cdn) {
			var row = locals[cdt][cdn];
			return {
				query: "golden.golden.stock.uom_query",
				filters: { 'item_code': row.item_code }
			}
		});
	},
	refresh: function(frm) {
		if(frm.doc.docstatus == 0 || frm.doc.__islocal) {
			frm.add_custom_button(__("Get Purchase Order"), function() {
				items = $.map( cur_frm.doc.items, function(item,idx) { return item.purchase_order } )
				added_items = items.join(",")
				erpnext.utils.map_current_doc({
					method: "golden.golden.doctype.receive_order.receive_order.get_purchase_order",
					source_doctype: "Purchase Order",
					target: frm,
					setters:  {
						company: frm.doc.company || undefined,
					},
					get_query_filters: {
						docstatus: 1,
						company: frm.doc.company,
						status: ["in", "To Receive and Bill, To Receive"],
						name: ["not in", added_items]
					}
				})
			});
		}
		if(frm.doc.docstatus == 0 || frm.doc.__islocal){
			frm.events.set_read_only(frm);
		}
	},
	set_posting_time: function(frm){
		frm.events.set_read_only(frm);
	},
	set_read_only: function(frm){
		if(frm.doc.set_posting_time == 1){
			frm.set_df_property("posting_date", "read_only", false);
			frm.set_df_property("posting_time", "read_only", false);
		}else{
			frm.set_df_property("posting_date", "read_only", true);
			frm.set_df_property("posting_time", "read_only", true);
		}
	},
});
frappe.ui.form.on('Receive Order Item', {
	purchase_order: function(doc, cdt, cdn){
		var d = locals[cdt][cdn];
		if(d.purchase_order){
			frappe.call({
				method: "golden.golden.doctype.receive_order.receive_order.get_po_detail",
				args:{
					po: d.purchase_order,
					item_code: d.item_code
				},
				callback: function (r) {
					if(r.message) {
						frappe.model.set_value(cdt, cdn, r.message);
					}
				}
			})
		}
	},
	item_code: function(doc, cdt, cdn) {
		var d = locals[cdt][cdn];
		if(d.item_code){
			frappe.call({
				method: "frappe.client.get",
				args: {
					doctype: "Item",
					filters:{
						name: d.item_code,
					}
				},
				callback: function (data) {
					frappe.model.set_value(cdt, cdn, "stock_uom", data.message.stock_uom);
				}
			})
			// frappe.call({
			// 	method: "frappe.client.get",
			// 	args: {
			// 		doctype: "Purchase Order Item",
			// 		filters:{
			// 			item_code: d.item_code,
			// 			parent: d.purchase_order
			// 		}
			// 	},
			// 	callback: function (data) {
			// 		var qty = flt(data.message.qty) - flt(data.message.received_qty);
			// 		frappe.model.set_value(cdt, cdn, "purchase_order", data.message.parent);
			// 		frappe.model.set_value(cdt, cdn, "po_detail", data.message.name);
			// 		frappe.model.set_value(cdt, cdn, "qty", qty);
			// 		frappe.model.set_value(cdt, cdn, "uom", data.message.uom);
			// 		frappe.model.set_value(cdt, cdn, "conversion_factor", data.message.conversion_factor);
			// 	}
			// })
		}else{
			frappe.model.set_value(cdt, cdn, "stock_uom", "");
			frappe.model.set_value(cdt, cdn, "uom", "");
			frappe.model.set_value(cdt, cdn, "conversion_factor", "1");
		}
	},
	uom: function(frm, cdt, cdn){
		var d = locals[cdt][cdn];
		if(d.uom){
			return frappe.call({
				method: "golden.golden.doctype.receive_order.receive_order.get_conversion_factor",
				args: {
					parent: d.item_code,
					uom: d.uom
				},
				callback: function (data) {
					if(data.message){
						frappe.model.set_value(cdt, cdn, data.message);
					}
				}
			})
		}
	}
})
// cur_frm.set_query("purchase_order", "items",  function (doc, cdt, cdn) {
// 	var c_doc= locals[cdt][cdn];
// 	return {
// 		query: "golden.golden.doctype.receive_order.receive_order.get_list_purchase_order",
// 		filters: {
// 			'item_code': c_doc.item_code,
// 		}
// 	}
// });
// cur_frm.set_query("item_code", "items",  function (doc, cdt, cdn) {
// 	var c_doc= locals[cdt][cdn];
// 	return {
// 		query: "golden.golden.doctype.receive_order.receive_order.get_item_code",
// 	}
// });
