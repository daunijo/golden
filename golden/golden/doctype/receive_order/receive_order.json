{
 "allow_copy": 0, 
 "allow_guest_to_view": 0, 
 "allow_import": 0, 
 "allow_rename": 0, 
 "autoname": "naming_series:", 
 "beta": 0, 
 "creation": "2018-02-15 11:02:24.464608", 
 "custom": 0, 
 "docstatus": 0, 
 "doctype": "DocType", 
 "document_type": "", 
 "editable_grid": 1, 
 "engine": "InnoDB", 
 "fields": [
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "sb01", 
   "fieldtype": "Section Break", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "default": "{container}", 
   "fieldname": "title", 
   "fieldtype": "Data", 
   "hidden": 1, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Title", 
   "length": 0, 
   "no_copy": 1, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "naming_series", 
   "fieldtype": "Select", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Series", 
   "length": 0, 
   "no_copy": 1, 
   "options": "RO-", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 1, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "container", 
   "fieldtype": "Data", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 1, 
   "in_standard_filter": 0, 
   "label": "Container", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "etd", 
   "fieldtype": "Date", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "ETD", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "eta", 
   "fieldtype": "Date", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "ETA", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "accepted_location", 
   "fieldtype": "Link", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Accepted Location", 
   "length": 0, 
   "no_copy": 0, 
   "options": "Warehouse", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "from_stock_reconciliation", 
   "fieldtype": "Check", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "From Stock Reconciliation", 
   "length": 0, 
   "no_copy": 1, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 1, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "cb01", 
   "fieldtype": "Column Break", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "expedition", 
   "fieldtype": "Link", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Expedition", 
   "length": 0, 
   "no_copy": 0, 
   "options": "Expedition", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "default": "Today", 
   "fieldname": "posting_date", 
   "fieldtype": "Date", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 1, 
   "in_standard_filter": 0, 
   "label": "Posting Date", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "posting_time", 
   "fieldtype": "Time", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 1, 
   "in_standard_filter": 0, 
   "label": "Posting Time", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "depends_on": "eval:doc.docstatus==0", 
   "fieldname": "set_posting_time", 
   "fieldtype": "Check", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Edit Posting Date and Time", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "company", 
   "fieldtype": "Link", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Company", 
   "length": 0, 
   "no_copy": 0, 
   "options": "Company", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "amended_from", 
   "fieldtype": "Link", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Amended From", 
   "length": 0, 
   "no_copy": 1, 
   "options": "Receive Order", 
   "permlevel": 0, 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 1, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "sb02", 
   "fieldtype": "Section Break", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "items", 
   "fieldtype": "Table", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Items", 
   "length": 0, 
   "no_copy": 0, 
   "options": "Receive Order Item", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 1, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "delivery", 
   "fieldtype": "Table", 
   "hidden": 1, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Delivery", 
   "length": 0, 
   "no_copy": 0, 
   "options": "Receive Order Item Delivery", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }
 ], 
 "has_web_view": 0, 
 "hide_heading": 0, 
 "hide_toolbar": 0, 
 "icon": "fa fa-truck", 
 "idx": 0, 
 "image_view": 0, 
 "in_create": 0, 
 "is_submittable": 1, 
 "issingle": 0, 
 "istable": 0, 
 "max_attachments": 0, 
 "modified": "2018-04-30 14:26:00.151845", 
 "modified_by": "Administrator", 
 "module": "Golden", 
 "name": "Receive Order", 
 "name_case": "", 
 "owner": "Administrator", 
 "permissions": [
  {
   "amend": 1, 
   "apply_user_permissions": 0, 
   "cancel": 1, 
   "create": 1, 
   "delete": 1, 
   "email": 1, 
   "export": 1, 
   "if_owner": 0, 
   "import": 0, 
   "permlevel": 0, 
   "print": 1, 
   "read": 1, 
   "report": 1, 
   "role": "System Manager", 
   "set_user_permissions": 0, 
   "share": 1, 
   "submit": 1, 
   "write": 1
  }, 
  {
   "amend": 1, 
   "apply_user_permissions": 0, 
   "cancel": 1, 
   "create": 1, 
   "delete": 1, 
   "email": 1, 
   "export": 1, 
   "if_owner": 0, 
   "import": 0, 
   "permlevel": 0, 
   "print": 1, 
   "read": 1, 
   "report": 1, 
   "role": "Administrator", 
   "set_user_permissions": 0, 
   "share": 1, 
   "submit": 1, 
   "write": 1
  }
 ], 
 "quick_entry": 0, 
 "read_only": 0, 
 "read_only_onload": 0, 
 "show_name_in_global_search": 0, 
 "sort_field": "modified", 
 "sort_order": "DESC", 
 "title_field": "title", 
 "track_changes": 1, 
 "track_seen": 0
}